### PR TITLE
Parse to Result<>, and Keyboard-based menu system

### PR DESCRIPTION
**Pull request by geralddejong**
There was a lot of unwrapping and quite a few cryptic non-pest error messages

### CODE REVIEW

File: src/application.rs [hunk 1]
```diff
@@ -151,24 +170,10 @@ impl Application {
...
-            VirtualKeyCode::Key3 => self.crucible.set_speed(25),
-            VirtualKeyCode::Key4 => self.crucible.set_speed(125),
-            VirtualKeyCode::Key5 => self.crucible.set_speed(625),
-            VirtualKeyCode::B => self.create_brick(),
-            VirtualKeyCode::F => self.scene.select_next_face(None, self.crucible.fabric()),
-            VirtualKeyCode::M => self.scene.watch_midpoint(),
-            VirtualKeyCode::O => self.scene.watch_origin(),
-            _ => {}
-        }
+        self.user_interface.key_pressed(keycode);
```
Comment (mitchelljustin) [inline src/application.rs:173]: Clean

File: src/build/mod.rs [hunk 2]
```diff
@@ -5,4 +5,4 @@ pub mod sphere;
 pub mod tenscript;
 pub mod example;
 pub mod brick;
-pub mod brick_proto;
+pub mod brick_proto;
```
Comment (mitchelljustin) [inline src/build/mod.rs:8]: Is there still a reason for this mod to exist? I would just fold it into `brick`

File: src/build/tenscript/library.scm [hunk 3]
```diff
@@ -1,69 +1,71 @@
...
-    (name "Flagellum")
+    (name "Simple" "Flagellum")
     (build
       (branch (alias Single)
-        (face (alias Top) (grow 20 (scale .9)))))
+        (face (alias :next-base) (grow 20 (scale .9)))))
     (shape))
   (fabric
-    (name "Halo by Crane")
+    (name "Art" "Halo by Crane")
```
Comment (mitchelljustin) [inline src/build/tenscript/library.scm:21]: Haha I like it

File: src/build/tenscript/mod.rs [hunk 4]
```diff
@@ -27,16 +27,33 @@ mod build_phase;
...
 #[derive(Debug)]
-pub enum ParseError {
+pub enum TenscriptError {
     Pest(Error<Rule>),
+    Format(String),
     Invalid(String),
 }
 
-impl Display for ParseError {
+impl TenscriptError {
```
Comment (mitchelljustin) [inline src/build/tenscript/mod.rs:36]: Weird to put these convenience methods inside an impl called `TenscriptError`.

File: src/build/tenscript/mod.rs [hunk 5]
```diff
@@ -180,20 +197,20 @@ impl Library {
         let source = fs::read_to_string("src/build/tenscript/library.scm").unwrap();
         match Self::from_tenscript(&source) {
             Ok(library) => library,
-            Err(ParseError::Pest(error)) => panic!("pest parse error: \n{error}"),
-            Err(e) => panic!("{e:?}")
+            Err(TenscriptError::Pest(_error)) => panic!("pest parse error: \n{_error}"),
+            Err(_error) => panic!("{_error:?}")
```
Comment (mitchelljustin) [inline src/build/tenscript/mod.rs:201]:
```suggestion
            Err(TenscriptError::Pest(error)) => panic!("pest parse error: \n{error}"),
            Err(error) => panic!("{error:?}")
```
Only prefix underscore if unused

File: src/build/tenscript/mod.rs [hunk 6]
```diff
@@ -27,16 +27,33 @@ mod build_phase;
...
 
-impl Display for ParseError {
+impl TenscriptError {
+    pub fn parse_float(string: &str, spot: &str) -> Result<f32, Self> {
+        string.parse().map_err(|_| TenscriptError::Format(format!("[{spot}]:Not a float: '{string}'")))
+    }
+
+    pub fn parse_usize(string: &str, spot: &str) -> Result<usize, Self> {
+        string.parse().map_err(|_| TenscriptError::Format(format!("[{spot}]Not an int: '{string}'")))
+    }
```
Comment (mitchelljustin) [inline src/build/tenscript/mod.rs:43]:
```suggestion
    pub fn parse_float(string: &str, spot: &str) -> Result<f32, Self> {
        string.parse().map_err(|_| TenscriptError::Format(format!("[{spot}]: Not a float: '{string}'")))
    }

    pub fn parse_usize(string: &str, spot: &str) -> Result<usize, Self> {
        string.parse().map_err(|_| TenscriptError::Format(format!("[{spot}]: Not an int: '{string}'")))
    }
```

File: src/user_interface/mod.rs [hunk 7]
```diff
@@ -31,8 +32,82 @@ pub enum Action {
...
     CalibrateStrain,
+    ToggleDebug,
+    SetSpeed(usize),
+    CreateBrick,
+    SelectNextFace,
+    WatchMidpoint,
+    WatchOrigin,
+}
+
+fn fabric_menu(fabrics: &Vec<FabricPlan>, below: Vec<String>) -> Vec<Menu> {
```
Comment (mitchelljustin) [inline src/user_interface/mod.rs:46]:
```suggestion
fn fabric_menu(fabrics: &[FabricPlan], below: Vec<String>) -> Vec<Menu> {
```

File: src/build/tenscript/mod.rs (hunk 4 above)
Comment (geralddejong) [reply src/build/tenscript/mod.rs:36]: Suggestion?

Comment (geralddejong) [reply src/build/tenscript/mod.rs:36]: Weird is not an argument.  What would be not weird?

**Review by mitchelljustin — APPROVED** [review]

**Pull request merged**